### PR TITLE
Zach log in page

### PR DESCRIPTION
**Pull request by ZachDFK**
Combining lasted branches together to see if we can make them fit together

### CODE REVIEW

File: app/views/sessions/home.html.erb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<h1>Yar a wizard and you're home!</h1>
+<p id='notice'> <%=notice%></p>
+
+<%= link_to 'Login', login_path %>
+<%= link_to 'Singup', signup_path %>
```
Comment (GoodOlChuck) [inline app/views/sessions/home.html.erb]: Small typo 'Singup'

File: app/views/users/_signup_form.html.erb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+      <ul>
+      <% @user.errors.full_messages.each do |message| %>
+        <li><%= message %></li>
+      <% end %>
+      </ul>
+    </div>
+  <% end %>
+
+<% end %>
```
Comment (GoodOlChuck) [inline app/views/users/_signup_form.html.erb]: There is no newline at the end of this file.

File: Gemfile [hunk 3]
```diff
@@ -12,6 +12,7 @@ gem 'sass-rails', '~> 5.0'
 gem 'therubyracer', platforms: :ruby
 gem 'turbolinks'
 gem 'uglifier', '>= 1.3.0'
+gem 'bcrypt'
```
Comment (kevincox) [inline Gemfile:15]: I still think that it is weird that you have to do this but whatever.

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -3,7 +3,7 @@ class ApplicationController < ActionController::Base
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
 
- def hello
-	render text:"hi bud"
- end
+ 	def hello
+		render text:"hi bud"
+	end
```
Comment (kevincox) [inline app/controllers/application_controller.rb]: I think you used tabs here while the rest of the file is spaces.

File: app/models/user.rb [hunk 5]
```diff
@@ -1,16 +1,45 @@
 class User < ActiveRecord::Base
-  has_many :posts
-  validates :password, presence: true, length: {minimum: 8}
-  validates :username, presence: true, length: {minimum: 3}
-  validates :email, uniqueness: true, presence: true
+has_many :posts
+has_many :comments
+require 'bcrypt'
```
Comment (kevincox) [inline app/models/user.rb]: Fix the indent please.

File: app/views/posts/_form.html.erb [hunk 6]
```diff
@@ -21,7 +21,7 @@
   </div>
   <div class="field">
     <%= f.label :author %><br>
-    <%= f.number_field :author %>
+    <%= f.label User.get_username_by_id(@post.author) %>
```
Comment (kevincox) [inline app/views/posts/_form.html.erb]: Is this temporary? Why do they need to specify the author in the form? Shouldn't it use the logged in user?

File: app/views/posts/index.html.erb [hunk 7]
```diff
@@ -17,7 +17,7 @@
       <tr>
         <td><%= post.title %></td>
         <td><%= post.contents %></td>
-        <td><%= post.author %></td>
+        <td><%= User.get_username_by_id(post.author) %></td>
```
Comment (kevincox) [inline app/views/posts/index.html.erb]: Shouldn't you be able to do `post.author.username` or something?

**Comment by kevincox** [comment]
Some more tests would be nice. For example a login and logout test.

File: app/models/user.rb [hunk 8]
```diff
@@ -1,16 +1,45 @@
...
+	has_secure_password	  
+
+	def self.get_username_by_id(id)
+		user = User.find_by_id(id)
+		if user
+			return user.username
+		else
+			return id
+		end
+	end
```
Comment (kevincox) [inline app/models/user.rb]: This method shouldn't be necessary. Something like `User.find_by_id(whatever).try :username` will work. I don't see the value in this abstraction.

File: app/models/user.rb [hunk 9]
```diff
@@ -1,16 +1,45 @@
...
+	def self.authenticate_with_username_or_email(username_or_email="",login_password="")
+		if EMAIL_REGEX.match(username_or_email)
+			user = User.find_by_email(username_or_email)
+		else
+			user = User.find_by_username(username_or_email)
+		end
+		if user && user.authenticate(login_password)
```
Comment (kevincox) [inline app/models/user.rb]: Style nitpick.

``` rb
if user && user.authenticate(login_password)
   user
end
```

If you want to be extra snazzy use try.

``` rb
if user.try :authenticate, login_password
   user
end
```

I can't wait for the elvis operator to ship.

``` rb
if user.?authenticate login_password
   user
end
```

File: app/models/user.rb [hunk 10]
```diff
@@ -1,16 +1,45 @@
...
+
+	def self.get_username_by_id(id)
+		user = User.find_by_id(id)
+		if user
+			return user.username
+		else
+			return id
+		end
+	end
+	def self.authenticate_with_username_or_email(username_or_email="",login_password="")
```
Comment (kevincox) [inline app/models/user.rb]: Why do these have default arguments? I can't see a single use case of not passing them in. If there is a use case we can add them later.

File: app/controllers/comments_controller.rb [hunk 11]
```diff
@@ -1,7 +1,9 @@
 class CommentsController < ApplicationController
+  
+
   def create
   	@post = Post.find(params[:post_id])
-
+  	
   	@comment = @post.comments.create(comment_params)
   	redirect_to post_path(@post)
```
Comment (kevincox) [inline app/controllers/comments_controller.rb]: You're mixing tabs and spaces in this file.

**Pull request closed without merging**